sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java

```diff
@@ -201,8 +206,15 @@ public void processElement(
           break;
       }
 
+      // TODO(sjvanrossum): https://github.com/apache/beam/issues/31828
+      // NOTE: Null and empty ordering keys are treated as equivalent.
       @Nullable String topic = dynamicTopicFn.apply(element);
-      K key = keyFunction.apply(ThreadLocalRandom.current().nextInt(numShards), topic);
+      @Nullable String orderingKey = message.getOrderingKey();
+      int shard =
+          Strings.isNullOrEmpty(orderingKey)
```

Comment:
can we use the new logic only if sink is configured to care about ordering keys to avoid changing the batching for existing cases where key is set?

Reply:
That makes sense, after reverting the strict validation on ordering keys requires a change here as well to retain the existing behavior.

Reply:
See response in the comment [below](https://github.com/apache/beam/pull/31608#discussion_r1779098644). I think it makes sense to revert this new logic

Reply:
it seems we may want to group by ordering key so that we can do better at sequencing publishing to the ordering key if that prevents pubsub ingestion issues.

However to fix the issue that this new sharding is used even if we are not publishing the ordering key, perhaps we should instead clear the ordering key in the verification dofn if we don't want to publish the ordering key? That could be done before we include the ordering key in message size validation etc which doesn't make sense if it will later be ignored.